modules/deriving/src/main/scala/shapeless3/deriving/kinds.scala

```diff
@@ -22,19 +22,11 @@ import scala.Tuple.Union
 import scala.compiletime.*
 import scala.compiletime.ops.int.S
 import scala.deriving.*
-import scala.util.NotGiven
 
 object K0:
-  type Kind[C, O] = C {
-    type Kind = K0.type
-    type MirroredType = O
-    type MirroredMonoType = O
-    type MirroredElemTypes <: Tuple
-  }
-
-  type Generic[O] = Kind[Mirror, O]
-  type ProductGeneric[O] = Kind[Mirror.Product, O]
-  type CoproductGeneric[O] = Kind[Mirror.Sum, O]
+  type Generic[O] = Mirror.Of[O] { type Kind = K0.type }
+  type ProductGeneric[O] = Mirror.ProductOf[O] { type Kind = K0.type }
+  type CoproductGeneric[O] = Mirror.SumOf[O] { type Kind = K0.type }
 
   def Generic[O](using gen: Generic[O]): gen.type = gen
   def ProductGeneric[O](using gen: ProductGeneric[O]): gen.type = gen
```

Comment:
So far I have only interacted with Shapeless 3 through the `ProductInstances`, so I didn't consider this in the original issue #218, but I suspect that these methods as well as `Generic.derive` (which I can not comment on due to Github's broken UI), will suffer from the same problem as `mkInstances`. If `Generic.derive` is supposed to be an entry point in the library like in Magnolia, then it will not work correctly with custom `Mirror` instances since `Generic` still retains the `type Kind` refinement. `def Generic[O](using gen: Generic[O]): gen.type = gen` could probably be changed here to `def Generic[O](using m: Refl[Mirror, O]): Generic[O] & m.type = m.asInstanceOf` (inlining may be necessary), but this would only help if the user is summoning `Generic[O]` through `def Generic`. If the user is summoning `Generic[O]` through implicit search via an implicit parameter, it will not find plain `Mirror` instances.  In this light, I feel like the implicit conversions are still the most robust solution, as they will work in every case automatically. It's pretty much what you'd end up writing to derive `Generic` if it was a new type and not just a refinement on `Mirror`.

Reply:
`Generic.derive` is one of the extension methods I mentioned that needs the implicit scope of `type Kind`. I will look into changing `def Generic`. But an implicit conversion will not help with the extension methods and with the implicit scope.

Reply:
Why not? Note that in Scala 3, implicit conversions are specifically givens of type `Conversion[?,?]`, so strictly speaking my suggestion is not a real implicit conversion, just a regular ol' derivation. `Conversion`s are not considered during implicit search iirc, but the given from #218 should work fine. 

This is the situation I'm talking about:
```scala
def myDerivation(using g: Generic[Foo]) {
    g.derive(...)
}
```
as opposed to
```scala
def myDerivation(using Mirror[Foo]) {
    Generic.derive(...)
}
```
which should work after changing `def Generic` to depend on `Mirror`.

I suspect that `def Generic` will require the same inlining and Aux tricks and end up looking identical to my example from #218, except as a def instead of given.

Reply:
Dunno, seems to work simply by changing the type signature. If we define implicit conversions, they would have to be imported which is poor UX and a bigger change than the type alias.

Reply:
With "changing the type signature" do you mean the signature of `myDerivation`? My point is that both `myDerivation` and `Foo` may be in different third-party libraries, so neither the definition of `myDerivation` nor `Mirror[Foo]` can easily be changed by the enduser. They would have to essentially wrap the `Mirror[Foo]` themselves with a given like I've shown in #218 _or_ manually provide the implicit parameter at every use-site and cast it: `given MyDerived = myDerivation(using summon[Mirror.Of[Foo]].asInstanceOf[Generic[Foo]])`. 

If the implicit conversions were in a companion object to `Generic`, wouldn't they be available automatically? I'm not so familiar with the implicit resolution rules. We could also leave the `Generic` parameter in `mkInstances`  and try to provide a hint with `@implicitNotFound` that directs users to import the implicit conversions if their type has a custom `Mirror` instance. The question then is how to incorporate the regular error message for non-case-classes.

Reply:
I meant changing the signature of `def Generic`. `type Generic` is just a type alias so it has no companion object. Maybe it would have been better to make it an opaque type but idk if that would be binary compatible.

Reply:
But if you only change `def Generic` it won't fix the `myDerivation` example?

Reply:
I'm not sure I understand, what is not fixed exactly?

Reply:
If I'm not mistaken this still won't work, even if you change `def Generic`:
```scala
def myDerivation(using g: Generic[Foo]) {
    g.derive(...)
}
```
If this is the derivation code in some third-party library, then the end-user can not change the signature of `myDerivation` and is out of luck (not to mention that even if they could, the standard error message would not be particularly helpful for them to figure that out).

Reply:
But an implicit conversion would not fix that either

Reply:
Why not? I don't see how `myDerivation` is different in that respect from `K0.mkProductInstances`, which I have confirmed to work. When `myDerivation` is called, it will look for givens of type `Generic[Foo]` in scope at the use site. It will then find the "implicit conversion" given from #218, look in turn for an instance of `Mirror.Of[T]` and find it. The conversion givens must be somehow imported, of course.

Reply:
No, implicit conversions don't work like that. I've added a normal given, also seems to work.

Reply:
But it isn't an implicit conversion, strictly speaking. Your given works exactly the same as mine, except it doesn't seem to need the inline and Aux stuff to remember the concrete Tuple types.

Reply:
Ok, I just got confused by the entire talk about implicit conversions. So you were suggesting to add a given. It makes sense now, this has nothing to do with implicit conversions. 

> Your given works exactly the same as mine, except it doesn't seem to need the inline and Aux stuff to remember the concrete Tuple types, unless the test is somehow not testing it sufficiently.

`m.type` makes sure the type is preserved.